File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/dataset/CustomerOrders.scala
package io.devorbit.sparklabs.batch.sales.dataset

import org.apache.spark.sql.functions.{approx_count_distinct, col, desc}

object CustomerOrders extends App {
  DatasetUtils.toDataFrame(DatasetUtils.customers)
    .join(DatasetUtils.toDataFrame(DatasetUtils.orders), Seq("customer_id"), "inner")
    .groupBy(col("customer_id"))
    .agg(approx_count_distinct("customer_id").as("total_orders"))
    .orderBy(desc("total_orders"))
    .show(false)
}

File: r/sparklabs/batch/sales/sql/customer_orders.R
source("r/sparklabs/batch/sales/sql/utils.R")

showDF(run_sql("
  SELECT c.customer_id, approx_count_distinct(c.customer_id) AS total_orders
  FROM customers c
  JOIN orders o ON c.customer_id = o.customer_id
  GROUP BY c.customer_id
  ORDER BY total_orders DESC
"), truncate = FALSE)
